xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/less/general.less

```diff
@@ -131,4 +131,81 @@ img {
   -moz-hyphens: auto;
   -webkit-hyphens: auto;
   hyphens: auto;
+}
+
+// User preferences ===========================================================
+body.content {
+  /* We apply various user preference related styles to the content of the page.
+   We use classes that are added to the body itself in htmlheader.vm */
+  // XWiki font size preference
+  &.preference-font-size-regular {
+    font-size: 140%;
```

Comment:
I would have left the font size unchanged in this case, also to avoid introducing any regressions. For example, I see that when setting the font size in the theme customization, this value is also applied to the body element. So when you set the font size, e.g., to `20px` in the theme customization, you'll have a `font-size: 20px` on `body`. This CSS would completely break this. Further, also all other font sizes are not relative to the font size set in the theme customization. When the font size in the theme is set to `20px`, all these settings apart from the `largest` setting would actually decrease the font size.

Reply:
> I would have left the font size unchanged in this case, also to avoid introducing any regressions. For example, I see that when setting the font size in the theme customization, this value is also applied to the body element. 

Okay, so I just checked, and bootstrap uses two default hard coded values at different levels :/
There's a `10px` on the `html` node, and `font-size: @font-size-base;` on the body node(with its value expected to be in px).
This makes little sense to me and IMO that's an issue with bootstrap 3.
The way things are now, the 10px on the html node is completely useless...

> So when you set the font size, e.g., to 20px in the theme customization, you'll have a font-size: 20px on body.

With the solution above, the theme customization value would be overriden by the percentage which has more specificity. 
Our font-sizes are defined in multiple ways:
* percentage, rems and ems: depend on the relative font size (native CSS)
* `0.9*@font-size-base` : depend on the bootstrap value.

The only point of setting `@font-size-base` on the body is to allow using relative units without breaking presentation.
So I guess the way I should implement this is with a percent on the bootstrap value... However I feel like even this solution would break some UIs somehow...

This is an overly complex solution and I dislike it a lot. I'll make a proposal to start uniformizing our code base font-sizes so that we don't get stuck in an interwoven web of dependencies going through ten style files and two fundamentally incompatible models next time we try to make a system change relative to font sizes.

I think the best solution for long term would be to override bootstrap code with something that makes sense with the modern CSS capabilities, this would provide code that's easier to understand and maintain on the long term.

---
In my opinion we should aim at:
* The only absolute units for font sizes are used on `:root` (or `html`).
* Every other font size should be defined using percentages , ems or rems

Reply:
The changes described here have been pushed on this PR in https://github.com/xwiki/xwiki-platform/pull/2694/commits/09af3b4516658f4e8b3356a9c64fce626a0d607b :+1:

Reply:
I absolutely agree with you that we need to get rid of the current font size system and use one that is based on either a single font size or even the browser's font size. However, the problem is that this font size that is set by bootstrap is not only used for computing other font sizes but also for paddings and margins (search for `@line-height-computed` in the source code).

However, this will be a lot of work and I think we should first decide if we want to stay with Bootstrap 3 or not before we put more work into modernizing Bootstrap 3.

Ideally, the font size settings shouldn't be part of this PR. However, I see the problem that you need to split the accessibility settings in order to offer several options for underlining links.

Reply:
> Ideally, the font size settings shouldn't be part of this PR. However, I see the problem that you need to split the accessibility settings in order to offer several options for underlining links.

As I wrote in my previous comment, maybe it would make sense to just drop the font size feature for now (requires a vote) and to simply replace the accessibility setting by the underline links setting.

Reply:
> However, the problem is that this font size that is set by bootstrap is not only used for computing other font sizes but also for paddings and margins (search for @line-height-computed in the source code).

Thank you for highlighting this, I did not connect the dots here. It will be a pain moving, since the basic semantics of this system look skewed. This is kind of similar to a regression we had earlier this month because we used a foreground color as a background :) .

> However, this will be a lot of work and I think we should first decide if we want to stay with Bootstrap 3 or not before we put more work into modernizing Bootstrap 3.

From what I understood, the only reason why we stay on bootstrap 3 is the cost of migration (without it I'm pretty sure we'd have moved to the latest bootstrap in an heartbeat :) ). My idea would be to be less reliant on bootstrap 3 for font-size, which would make it easier to move to another system later.

> Ideally, the font size settings shouldn't be part of this PR. However, I see the problem that you need to split the accessibility settings in order to offer several options for underlining links.

:+1: It's a bit of a pain but it's something I would have added not long after anyways.

Reply:
> I would have left the font size unchanged in this case, also to avoid introducing any regressions. For example, I see that when setting the font size in the theme customization, this value is also applied to the body element. So when you set the font size, e.g., to `20px` in the theme customization, you'll have a `font-size: 20px` on `body`. This CSS would completely break this. Further, also all other font sizes are not relative to the font size set in the theme customization. When the font size in the theme is set to `20px`, all these settings apart from the `largest` setting would actually decrease the font size.

By the way I just realized that the legacy way to implement the `accessible` mode also overrode the `@font-size-base` defined in the colortheme and bootstrap :)

Reply:
My comment here was only regarding the "regular" size, "this case" meant the default "regular" font size. My worry here was primarily that the configured font size should be respected and left unchanged as much as possible when the font size setting is on the default value.

Reply:
The point that the accessible font size could end up smaller than the regular one was already a thing though